closures based client & more

mruzicka opened the pull request:
This was supposed to be a really small PR - I just wanted to add JSON encoding for the ```counts_filter``` parameter (there are in total 3 parameters which need JSON encoding: ```query```, ```order_by```, ```counts_filter```) - but ended up quite big including turning the client map object into a function reference, which is further complicated by the fact that I introduce a kind of polymorphism as known in OO programing with that function.

scottyw commented:
I've got slightly mixed feelings. It seems like a major re-architecture of the project.

On one hand, even a cursory glance tells me it's much better and past experience tells me it's probably rock solid.

On the other, it brings a lot of risk to stuff we've already tested. It must also have been a ton of effort for you and will be lots of effort to review.

I'd love to see this be the future but I'm wary of such big changes at this stage. Would you consider making this the first commit of a 2.0 release? I think we could add support for counts_filter in the same hacky way I did order_by in relative few lines in core.clj for now. Then we bring clj-puppetdb 2.0 in first thing on SGnext.

What are your feelings? Perhaps I am being too conservative but the amount of red and green makes me nervous.

mruzicka commented:
What about scaling it down, dropping the closures based client feature (which I think is the most controversial one) but keeping the unified handling of the parameters which need JSON encoding (and using that consistently even from the ```lazy-query``` which bypasses the the ```order-by```/```order_by``` encoding currently)?

scottyw commented:
That sounds like a reasonable approach though we should maybe take whichever option has least cost in engineering terms. If there's a clear delineation between the client and JSON encoding such that your proposal can be implemented by reverting a few files, then that seems a good plan to me. If there's hours of work to do that split then I'd rather take everything now. 

Getting review input from @kbrezina or @dmcpl seems useful too. I'll add some more detailed comments today too.

scottyw commented:
Let's take it all but get more eyes on it. That work for you?
 On 2 Apr 2015 10:24, "mruzicka" <notifications@github.com> wrote:

> While I'd prefer to get it in in its entirety, the effort to scale it back
> is not prohibiting either - my estimate is 2 hours.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/puppetlabs/clj-puppetdb/pull/11#issuecomment-88838686>
> .
>

mruzicka commented:
@scottyw The more people see it the better.

kbrezina commented on src/clj_puppetdb/http.clj:
```diff
@@ -24,47 +24,86 @@
+                            q/params->json
+                            map->query))))
+         ; now call the single argument variant of the function to execute the query
+         (this (.toString query))))
```
StringBuilder doesn't seem to be very Clojury. The following code might be more proper:

    (this (apply str (concat host
                             path
                             (if-not (empty? params)
                               (concat "?"
                                       (-> params
                                           q/params->json
                                           map->query))))))

mruzicka replied:
The intent here is optimization. While admittedly not very clojury, it is as straightforward as reasonably possible. Moreover the ```(str)``` function uses internally ```StringBuilder``` too.

scottyw replied:
Is there any reason to think clojure.string/join is less performant?
Arguably we shouldn't optimize for speed above all else anyway. I think
there's value in idiomatic code and simplicity even at the cost of some
degree of outright speed.

On 2 April 2015 at 14:15, mruzicka <notifications@github.com> wrote:

> In src/clj_puppetdb/http.clj
> <https://github.com/puppetlabs/clj-puppetdb/pull/11#discussion_r27655014>:
>
> > +      ; This arity-overloaded variant of the function is the one called by the `GET` function.
> > +      ; The `this` parameter refers under normal circumstances to this very function and it is
> > +      ; needed to be able to call another arity-overloaded variant of the function from the
> > +      ; function body. When the `this` parameter refers to a different function then it
> > +      ; effectively overrides the other arity-overloaded variants of the function.
> > +      ([this path params]
> > +       (let [query (-> (StringBuilder. host)
> > +                       (.append path))]
> > +         (if-not (empty? params)
> > +           (-> query
> > +               (.append \?)
> > +               (.append (-> params
> > +                            q/params->json
> > +                            map->query))))
> > +         ; now call the single argument variant of the function to execute the query
> > +         (this (.toString query))))
>
> The intent here is optimization. While admittedly not very clojury, it is
> as straightforward as reasonably possible. Moreover the (str) function
> uses internally StringBuilder too.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/puppetlabs/clj-puppetdb/pull/11/files#r27655014>.
>

mruzicka replied:
The performance of the those two is most probably the same under the assumption they are called comparable number of times. (The original code called them at least twice if not even greater number of times and that is what I was trying to avoid. Given that especially the query can be quite long calling ```(str host path (if query (str \? query)))``` seemed like a waste.)
If you

scottyw replied:
We're in philosophical territory here but my preference is towards the more "Clojurey" style that Karel alluded to. I think succinctness and simplicity are some of the advantages of Clojure and we risk losing them when transliterating Java. Certainly that'll sometimes be necessary but I'd reserve it for the occasions we know we measurable problems - like the slow reports endpoint thing we've seen this week for example. Doing things the "Clojure way" should also let us take advantage of improved tooling and compilers for free over the coming months/years.

Like I say, philosophical territory more than anything.

mruzicka replied:
OK, let's go with pure clojure then.

kbrezina commented on src/clj_puppetdb/http.clj:
```diff
@@ -24,47 +24,86 @@
+      ; This arity-overloaded variant of the function is the one which actually executes
+      ; the PDB query.
+      ([query]
+        #_(println "GET:" query)                                 ;; uncomment this to watch queries
```
IMHO, it should be possible to use standard logging mechanism with different log levels. This would probably be in DEBUG log level.

Take a look at pe-rbac-service for inspiration.

mruzicka replied:
That makes a lot of sense.

kbrezina commented:
- IMHO, since we'll do a lot of testing with PDB3 we can still do significant changes.
- Clojure purists probably wouldn't be very happy with the OO approach, however, I like the VCR separation.

kbrezina commented on src/clj_puppetdb/http.clj:
```diff
@@ -112,26 +153,21 @@
+    (let [query-info (-> (client)                           ; get client info
+                         (assoc :endpoint path)
+                         (assoc :params   params))
+          response (-> (client client path params)
```
This looks quite weird
    (client client path params)

However, it's probably not worth the effort to make it nicer. We've already discussed it with @mruzicka

scottyw replied:
As per above, naming the anonymous function in make-client-common removes the need to pass that first argument.

mruzicka replied:
Unfortunately this then prohibits the overriding from vcr enabled client.
The vcr enabled client calls the standard client while passing itself in the ```this``` argument thus making the standard client call the vcr client's version of the of the different arity overloaded variants of the client function.
The sequence of calls is then as follows:
1. the GET method calls the vcr client's [this path params] arity overloaded variant passing the vcr client in the ```this``` parameter
2. the vcr client's [this path params] sorts the known nested structures in params and calls the standard client's [this path params] arity overloaded variant passing on the ```this``` parameter unchanged (i.e. pointing to the vcr client)
3. the standard client's [this path params] forms the PDB query URL by converting the params into a query string and concatenating it to the ```host``` and ```path``` and uses the ```this``` parameter to call the [query] arity overloaded variant - which means (given that ```this``` points to the vcr client) that it calls the vcr client's version of that arity overloaded variant in this case
4. vcr client's [query] arity overloaded variant first tries to look up a response for the query in the ```vcr-dir``` and when that fails it calls the standard client's version of the [query] arity overloaded variant to actually query the PDB after which the vcr client's [query] arity overloaded variant saves the response for future use

scottyw replied:
OK, I hadn't got that. I don't love the OO thing but I haven't a better proposal either. I'm repeating what Karel said anyway and you guys have talked it over so just disregard my comment.

kbrezina commented on src/clj_puppetdb/http.clj:
```diff
@@ -24,47 +24,86 @@
+      ; needed to be able to call another arity-overloaded variant of the function from the
+      ; function body. When the `this` parameter refers to a different function then it
+      ; effectively overrides the other arity-overloaded variants of the function.
+      ([this path params]
```
I'd probably replace `this` with `handler`. In this context `this` usually means `this function` but sometimes even `that function`.

mruzicka replied:
The name ```this``` was inspired by java, meaning 'this object' - in this instance the client function called by the ```clj-puppetdb.http/GET``` function.

scottyw replied:
You can name an anonymous function i.e.

(fn xxyyzz

The recursive call becomes:

(xxyyzz (.toString query))))

Then there is no need to pass "this" in at all, which makes this clearer to my eyes, and also resolves Karel's comment about (client client...) below.

mruzicka replied:
Since (as explained above) we need the [this path params] arity overloaded variant to make the overriding possible the best we can do is to define yet another arity overloaded variant like so:
```clojure
    (fn stadarad_client
      ([path params]
        (standard_client standard_client path params))
      ...
```
we would have to add a similar variant to the vcr client as well:
```clojure
    (fn vcr_client
      ([path params]
        (vcr_client vcr_client path params))
      ...
```
With this in place we could call the client from the ```GET``` function simply as:
```clojure
  (client path params)
```
But I consider this a cosmetic change which unnecessarily adds another function call.

kbrezina commented:
I've commented a few minor thinks. However, it looks good to me. Good job. :thumbsup:

scottyw commented on src/clj_puppetdb/http.clj:
```diff
@@ -24,47 +24,86 @@
   `puppetlabs.ssl-utils.core/pems->ssl-context` function."
   [:ssl-cert :ssl-key :ssl-ca-cert])
 
+(defn- make-client-common
```
I really like this approach to the client.

+1 to Karel's comment on vcr separation. Way better than before.

scottyw commented on src/clj_puppetdb/vcr.clj:
```diff
@@ -22,21 +21,29 @@
   "Turn response body into a string."
   (let [charset (util/response-charset response)]
-    (update-in response [:body] #(slurp (BufferedReader. (InputStreamReader. % charset))))))
+    (->> (-> (get response :body)
```
I'm a broken record on this but I'd slurp and spit here to simplify the code. Ultimate VCR performance isn't a concern.

mruzicka replied:
I'm afraid we have to respect the encoding in the message header here, otherwise we might be misinterpreting the data.
But we can drop similar construct in the vcr.clj where I quite arbitrarily introduced the:
```clojure
(def vcr-charset (Charset/forName "UTF-8"))
```
I'll drop that.

mruzicka replied:
Just a note: We can't use
```clojure
(slurp % :ecoding charset)
```
as ```charset``` is an instance of ```java.nio.charset.Charset``` while the argument to ```:encoding``` must be a string.

scottyw commented on src/clj_puppetdb/http.clj:
```diff
@@ -1,11 +1,12 @@
             [clj-puppetdb.http-util :as util]
+            [clj-puppetdb.query :as q]
             [clj-puppetdb.schema :refer [Client]]
             [clj-puppetdb.util :refer [file?]]
```
raynes/fs already provides this method and we already have a dependency on it. We should delete that util namespace.

scottyw commented:
I've fairly got the boot in over this PR but I wanted to say explicitly that I think it's a great improvement. We've evolved this lib to breaking point and you've closed off a lot of grimy edge cases. Thanks for doing it.

Would really like the see the anonymous function in client named as described above, and think we should drop the "util" namespace. Apart from that, looks good to me!

mruzicka commented:
@scottyw @kbrezina I've added another commit to this PR in which I'm addressing your comments.
I deliberately added a separate commit to make it easier for you to see the changes. Could you please review it once again?
But please hold off the merge - I'd like to squash the second commit first.
Thanks for your comments!

kbrezina commented:
@mruzicka Looks good! Go ahead and squash it.

mruzicka commented:
@scottyw @kbrezina I've squashed the commits and added a description. It's now ready to be merged.
